lib/redis_mutex/lock.ex

```diff
@@ -1,118 +1,131 @@
-      use RedisMutex
-      alias PossumLodge.{Repo, Member}
-
-      def add_member(params) do
-        with_lock(params.phone_number) do
-          %Member{}
-          |> Member.changeset(params)
-          |> Repo.insert_or_update!
-        end
-      end
+if Code.ensure_loaded?(Redix) do
+  defmodule RedisMutex.Lock do
+    @moduledoc """
+    Defines the lock behaviour for RedisMutex.
+    """
+
+    @default_timeout :timer.seconds(40)
+    @default_expiry :timer.seconds(20)
+    @unlock_script """
+    if redis.call("get", KEYS[1]) == ARGV[1] then
+      return redis.call("del", KEYS[1])
+    else
+      return 0
     end
-    ```
-  """
-  defmacro with_lock(key, timeout \\ @default_timeout, expiry \\ @default_expiry, do: clause) do
-    quote do
-      key = unquote(key)
-      timeout = unquote(timeout)
-      expiry = unquote(expiry)
-      uuid = Uniq.UUID.uuid1()
+    """
+
+    @name RedisMutex
 
-      RedisMutex.Lock.take_lock(key, uuid, timeout, expiry)
+    @callback child_spec(opts :: Keyword.t()) :: Supervisor.child_spec()
 
-      block_value = unquote(clause)
+    @callback start_link(start_options :: RedisMutex.start_options()) ::
+                {:ok, pid()} | {:error, any()}
 
-      RedisMutex.Lock.unlock(key, uuid)
+    @callback with_lock(key :: String.t(), do: clause :: term()) :: any()
 
-      block_value
+    @callback with_lock(key :: String.t(), timeout :: integer(), do: clause :: term()) :: any()
+
+    @callback with_lock(key :: String.t(), timeout :: integer(), expiry :: integer(),
+                do: clause :: term()
+              ) ::
+                any()
+
+    @spec child_spec(Keyword.t()) :: Supervisor.child_spec()
+    def child_spec(opts) do
+      %{
+        id: __MODULE__,
+        start: {__MODULE__, :start_link, opts},
+        type: :worker,
+        restart: :permanent,
+        shutdown: 500
+      }
     end
-  end
 
-  @doc """
-  This function takes in a key, unique string, and a timeout in milliseconds.
-  It will call itself recursively until it is able to set a lock
-  or the timeout expires.
-  """
-  def take_lock(key, uuid, timeout \\ @default_timeout, expiry \\ @default_expiry, finish \\ nil)
+    @spec start_link(RedisMutex.start_options()) :: :ignore | {:error, any} | {:ok, pid}
+    def start_link(start_options \\ []) do
+      {redis_url, redix_opts} = Keyword.pop(start_options, :redis_url)
 
-  def take_lock(key, uuid, timeout, expiry, nil) do
-    finish = DateTime.add(DateTime.utc_now(), timeout, :millisecond)
-    take_lock(key, uuid, timeout, expiry, finish)
-  end
+      case redis_url do
+        redis_url when is_binary(redis_url) ->
+          Redix.start_link(redis_url, name: @name, sync_connect: true)
 
-  def take_lock(key, uuid, timeout, expiry, finish) do
-    if DateTime.compare(finish, DateTime.utc_now()) == :lt do
-      raise RedisMutex.Error, message: "Unable to obtain lock."
+        nil ->
+          [name: @name, sync_connect: true]
```

Comment:
I haven't been sure where to start this thread, but where we're setting a static name here it seems as good a place as any:

What is the objective of moving the process lifecycle out of redis_mutex's supervision tree for dependent apps to include in their own? Normally I would expect to see that in eliminating a singleton process, but we're still creating a singleton here. Since we're setting a static name here, there can only ever be one instance of this worker running. If someone wanted to run this on multiple redis clusters for some reason, how would they do that?

What if instead you let folks pass their own name on startup and to "with_lock"? I'm envisioning something like this:
```elixir
# lib/my_app/application.ex
[
  {RedisMutex, name: MyApp.Mutex},
  MyApp.Repo,
  MyAppWeb.Endpoint
]
```
```elixir
# lib/my_app/my_service.ex
with_lock MyApp.Mutex, "my_key", fn ->
  do_something_with_lock()
end
```
And perhaps we merge in config off the given name if folks don't want to pass it into the child spec?
```elixir
config :my_app, MyApp.Mutex, url: "localhost:6379"
```
And maybe an example in the docs for wrapping that in a package more similar to the singleton?
```elixir
defmodule MyApp.Mutex do
  def child_spec(opts) do
    opts
    |> Keyword.put(:name, __MODULE__)
    |> RedisMutex.child_spec()
  end
  
  def with_lock(key, opts \\ [], fun) do
    RedisMutex.with_lock(__MODULE__, key, opts, fun)
  end
end
```

Reply:
Thanks for the thoughts. Let me experiment with an approach like this and see how it goes.

While working through this, I tried an approach with wrapping `RedisMutex` in a a module. I had trouble trying to do that with a `with_lock` macro. Maybe if `with_lock` were a function instead, it would work better.

Reply:
@brettbeatty 

Thanks for the suggestions. I've made changes based on this comment. In particular, I've made updates that allows the user to pass in a name to the child spec and to `with_lock`. One thing this does is that it makes it a lot easier to re-use an existing named connection to Redis. In that case, there's no need to add anything new to the supervision tree. If an application already has a connection set up with `Redix`, then passing the name of that connection to the function is all that should be needed.